### ISSUE TITLE
Search by annotation target

### ISSUE THREAD

**Issue by tomcrane** [dlcs/elucidate-server#6]
Web annotation protocol allows additional query parameters, as described in https://github.com/IIIF/iiif.io/issues/951

Elucidate should support ?target=uri as an extension.

For consideration - how do we handle media fragments?

If I ask for annotations where `?target=http://example.org/iiif/my-canvas`, and Elucidate has the following annotation targets:

```
http://example.org/iiif/my-canvas#xywh=10,20,30,40
http://example.org/iiif/my-canvas#xywh=1000,1000,1000,1000
http://example.org/iiif/my-canvas
http://example.org/iiif/my-canvas#t=0,10
http://example.org/iiif/my-canvas#xywh=100,100,100,100&t=10,20
```

Should we only return `http://example.org/iiif/my-canvas` or do all of them match?

And if all of them match, how do we ask for annotations within a specific box/timespan/boxed-timespan?

Previous draft of IIIF Search API had a "box" param for annotations intersecting the region, but dropped for v1.0 to see what might happen with AV.

**Comment by tomcrane** [comment]
Summary of Slack discussion

* There is no W3C Annotation Search API
* But IIIF Content Search API is an effective guide for what that might look like
* So if we implemented a Search API on our Anno Server that felt like the IIIF Content Search API, with a W3C flavour, we'd be going in the right direction
* There's nothing wrong with Elucidate and Mathmos (Search Server) both offering subsets of the same Annotation Search API (and IIIF Content Search API explicitly), and nothing wrong with overlap - for some queries, you could ask either and get the same results. If you want complex full text operations over collections at scale, with understanding of "within" containment in a complex graph, you need Mathmos on top of Elucidate. To run a crowd sourcing project without complex full text search, you just need Elucidate.
* IIIF Content Search API is fine as a model - there's not really much about images in there!

?target=xxx - Elucidate definitely supports that
?q=some phrase - Mathmos definitely supports that

...and so on.

**Comment by danielgrant** [comment]
@tomcrane, @mattmcgrattan,

Current initial cut of `target` based search in Elucidate exists in the [`feature-search`](https://github.com/dlcs/elucidate-server/tree/feature-search) branch and builds on top of the existing `AnnotationCollection` and `AnnotationPage` types from the W3C Specification.

**Searching**
`GET http://annotation-local.digtest.co.uk/annotation/w3c/search?target=http://www.example.com/index1.html` ==> https://gist.github.com/danielgrant/6ee8b6c4b603a21672a5632b8f4d5b4d

**Paging Search Results**
`GET http://annotation-local.digtest.co.uk/annotation/w3c/search?page=0&target=http://www.example.com/index1.html&desc=1` ==> https://gist.github.com/danielgrant/9c2e69bb83145e0bb3833f7657c9a5fc

Collections and paging supports the container representation preferences as per [4.2.1](https://www.w3.org/TR/annotation-protocol/#container-representation-preferences) of the W3C Web Annotation Protocol. Paging also supports the `?iris=` and `?desc=` parameters.

**Comment by danielgrant** [comment]
> For consideration - how do we handle media fragments?

The current implementation only supports "strict" matching of targets, i.e. you would need to include the fragment in the `?target=` parameter to match the fragment.

I will investigate whether PostgreSQL JSON operators and functions support SQL-style `%` matchers.

**Comment by danielgrant** [comment]
Clients can now specify the `strict` parameter when executing searches.

When `true`, only exact matches to the provided `target` IRI will be returned.

When `false`, the `target` IRI is used as a "prefix" to matching only. This is useful for matching a `target` IRI that has a fragments suffixed to it, and the client wishes to match against these suffixed IRI's by simply providing the base IRI.

Note however that this isn't used for matching an IRI that only has fragments, it will match only IRI that has the provided `target` IRI as a prefix. For example `?target=http://www.example.com/index.htm&strict=false` will match both `http://www.example.com/index.htm` **and** `http://www.example.com/index.html`.

This may not be the desired functionality, in which case a more sophisticated implementation of the `annotation_search_by_target` PostgreSQL function will be required.

**Comment by danielgrant** [comment]
As of https://github.com/dlcs/elucidate-server/commit/31093cc6cff3ae404d8e72030c921d0691ca11ed, clients can now specify an optional `&box=x,y,w,h` parameter to the search URL that will return only `target`'s which intersect the specified box.

At present, this logic takes place in the `annotation_search` function of the database.

**Comment by danielgrant** [comment]
In order to support more sophisticated searching, an attempt has been made in the `feature-search-enhanced` branch to split certain parts of an Annotation into its constituent parts and store these in separate tables and columns in the database. In theory, this should allow better querying across these constituent parts.

In contrast to this, development will continue for now in the `feature-search` branch to support querying for Annoations by `body`, `target` with media fragment selectors without these Annotation splitting.

**Comment by danielgrant** [comment]
Closed as per #18.
